File: desktop-mend/src/main/java/co/samco/mend4/desktop/commands/StatePrinter.java
package co.samco.mend4.desktop.commands;

import co.samco.mend4.core.AppProperties;
import co.samco.mend4.core.CorruptSettingsException;
import co.samco.mend4.core.OSDao;
import co.samco.mend4.core.Settings;
import co.samco.mend4.desktop.core.I18N;
import co.samco.mend4.desktop.helper.SettingsHelper;
import co.samco.mend4.desktop.output.PrintStreamProvider;

import javax.inject.Inject;
import java.io.File;
import java.io.IOException;
import java.util.Arrays;
import java.util.List;
import java.util.Optional;
import java.util.function.Function;
import java.util.stream.Collectors;

public class StatePrinter extends Command {
    public static final String COMMAND_NAME = "get";
    public static final String ALL_FLAG = "-a";
    public static final String LOGS_FLAG = "-l";
    public static final String ENCS_FLAG = "-e";

    private final I18N strings;
    private final PrintStreamProvider log;
    private final OSDao osDao;
    private final SettingsHelper settingsHelper;
    private final Settings settings;

    private String arg;

    private final List<Function<List<String>, List<String>>> behaviourChain = Arrays.asList(
            a -> getArg(a),
            a -> checkArgIsFlag(a),
            a -> checkArgIsSetting(a),
            a -> fallbackUnknown()
    );

    @Inject
    public StatePrinter(I18N strings, PrintStreamProvider log, OSDao osDao, SettingsHelper settingsHelper,
                        Settings settings) {
        this.strings = strings;
        this.log = log;
        this.osDao = osDao;
        this.settingsHelper = settingsHelper;
        this.settings = settings;
    }

    private List<String> getArg(List<String> args) {
        if (args.size() != 1) {
            log.err().println(strings.getf("General.invalidArgNum", COMMAND_NAME));
            return null;
        }
        arg = args.get(0);
        return args;
    }

    private List<String> checkArgIsFlag(List<String> args) {
        try {
            if (arg.equals(ALL_FLAG)) {
                printAll();
                return null;
            } else if (arg.equals(LOGS_FLAG)) {
                printLogs();
                return null;
            } else if (arg.equals(ENCS_FLAG)) {
                printEncs();
                return null;
            }
        } catch (IOException | CorruptSettingsException e) {
            log.err().println(e.getMessage());
        }
        return args;
    }

    private List<String> checkArgIsSetting(List<String> args) {
        Optional<String> value = Arrays.stream(Settings.Name.values())
                .filter(n -> n.toString().equals(arg))
                .map(n -> settingsHelper.getSettingValueWrapped(n))
                .findFirst();

        if (value.isPresent()) {
            log.out().println(value.get());
            return null;
        } else return args;
    }

    private List<String> fallbackUnknown() {
        log.err().println(strings.getf("StatePrinter.settingNotFound", arg));
        return null;
    }

    private void printEncs() throws IOException, CorruptSettingsException {
        File encDir = new File(settings.getValue(Settings.Name.ENCDIR));
        String encs = Arrays.stream(osDao.getDirectoryListing(encDir))
                .filter(f -> osDao.getFileExtension(f).equals(AppProperties.ENC_FILE_EXTENSION))
                .map(f -> osDao.getBaseName(f))
                .collect(Collectors.joining(strings.getNewLine()));
        log.out().println(encs);
    }

    private void printLogs() throws CorruptSettingsException, IOException {
        File logDir = new File(settings.getValue(Settings.Name.LOGDIR));
        String logs = Arrays.stream(osDao.getDirectoryListing(logDir))
                .filter(f -> osDao.getFileExtension(f).equals(AppProperties.LOG_FILE_EXTENSION))
                .map(f -> osDao.getBaseName(f))
                .collect(Collectors.joining(strings.getNewLine()));
        log.out().println(logs);
    }

    private void printAll() {
        String output = Arrays.stream(Settings.Name.values())
                .map(n -> formatSettingValue(n))
                .collect(Collectors.joining(strings.getNewLine(2)));
        log.out().println(output);
    }

    private String formatSettingValue(Settings.Name name) {
        StringBuilder sb = new StringBuilder();
        sb.append(name.toString());
        sb.append("\t");
        String value = settingsHelper.getSettingValueWrapped(name);
        sb.append(value == null ? strings.get("StatePrinter.notFound") : value);
        return sb.toString();
    }

    @Override
    public void execute(List<String> args) {
        executeBehaviourChain(behaviourChain, args);
    }

    @Override
    public String getUsageText() {
        return strings.getf("StatePrinter.usage", COMMAND_NAME, ALL_FLAG, LOGS_FLAG, ENCS_FLAG,
                ALL_FLAG, LOGS_FLAG, ENCS_FLAG, settingsHelper.getSettingDescriptions());
    }

    @Override
    public String getDescriptionText() {
        return strings.get("StatePrinter.description");
    }

    @Override
    protected List<String> getCommandAliases() {
        return Arrays.asList(COMMAND_NAME);
    }

}

File: desktop-mend/src/main/java/co/samco/mend4/desktop/commands/StatePrinter.kt
package co.samco.mend4.desktop.commands

import co.samco.mend4.core.AppProperties
import co.samco.mend4.core.Settings
import co.samco.mend4.core.exception.CorruptSettingsException
import co.samco.mend4.desktop.core.I18N
import co.samco.mend4.desktop.dao.OSDao
import co.samco.mend4.desktop.dao.SettingsDao
import co.samco.mend4.desktop.helper.FileResolveHelper
import co.samco.mend4.desktop.helper.SettingsHelper
import co.samco.mend4.desktop.output.PrintStreamProvider
import org.apache.commons.io.FilenameUtils
import java.io.File
import java.io.IOException
import java.util.function.Function
import javax.inject.Inject

class StatePrinter @Inject constructor(
    private val strings: I18N,
    private val log: PrintStreamProvider,
    private val settingsHelper: SettingsHelper,
    private val settings: SettingsDao,
    private val fileResolveHelper: FileResolveHelper,
    private val osDao: OSDao
) : CommandBase() {
    private lateinit var arg: String

    private val behaviourChain = listOf(
        Function { a: List<String> -> getArg(a) },
        Function { a: List<String> -> checkArgIsFlag(a) },
        Function { a: List<String> -> printFromArg(a) },
    )

    private fun getArg(args: List<String>): List<String>? {
        if (args.size != 1) {
            log.err().println(strings.getf("General.invalidArgNum", COMMAND_NAME))
            return null
        }
        arg = args[0]
        return args
    }

    private fun checkArgIsFlag(args: List<String>): List<String>? {
        try {
            when (arg) {
                ALL_FLAG -> printSettings()
                LOGS_FLAG -> printLogs()
                ENCS_FLAG -> printEncs()
                else -> return args
            }
        } catch (t: Throwable) {
            failWithMessage(log, t.message)
        }
        return null
    }

    private fun printFromArg(args: List<String>): List<String>? {
        val value = SettingsDao.ALL_SETTINGS
            .filter { it.encodedName == arg }
            .map { settingsHelper.getSettingValueWrapped(it) }
            .firstOrNull()

        if (value != null) log.out().println(value)
        else log.err().println(strings.getf("StatePrinter.settingNotFound", arg))

        return null
    }

    private fun printEncs() {
        val encDir = settings.getValue(SettingsDao.ENC_DIR)
            ?.let { File(it) }
            ?: throw throw CorruptSettingsException(
                strings.getf(
                    "General.dirRequired",
                    SettingsDao.CURRENT_LOG,
                    COMMAND_NAME
                )
            )
        fileResolveHelper.assertDirectoryExists(encDir)

        osDao.listFiles(encDir)
            ?.filter { FilenameUtils.getExtension(it.name) == AppProperties.ENC_FILE_EXTENSION }
            ?.forEach { log.out().println(FilenameUtils.getBaseName(it.absolutePath)) }
    }

    @Throws(CorruptSettingsException::class, IOException::class)
    private fun printLogs() {
        val logDir = settings.getValue(SettingsDao.LOG_DIR)
            ?.let { File(it) }
            ?: throw CorruptSettingsException(
                strings.getf(
                    "General.dirRequired",
                    SettingsDao.LOG_DIR,
                    COMMAND_NAME
                )
            )

        fileResolveHelper.assertDirectoryExists(logDir)

        osDao.listFiles(logDir)
            ?.filter { FilenameUtils.getExtension(it.name) == AppProperties.LOG_FILE_EXTENSION }
            ?.forEach { log.out().println(FilenameUtils.getBaseName(it.absolutePath)) }
    }

    private fun printSettings() {
        SettingsDao.ALL_SETTINGS.forEach { log.out().println(formatSettingValue(it)) }
    }

    private fun formatSettingValue(name: Settings.Name): String {
        val value: String = settingsHelper.getSettingValueWrapped(name)
        val sb = StringBuilder()
        sb.append(name.toString())
        sb.append("\t")
        sb.append(value)
        return sb.toString()
    }

    public override fun execute(args: List<String>) {
        executeBehaviourChain(behaviourChain, args)
    }

    override val usageText: String
        get() = strings.getf(
            "StatePrinter.usage",
            COMMAND_NAME,
            ALL_FLAG,
            LOGS_FLAG,
            ENCS_FLAG,
            ALL_FLAG,
            LOGS_FLAG,
            ENCS_FLAG,
            settingsHelper.settingDescriptions
        )
    override val descriptionText: String
        get() = strings["StatePrinter.description"]
    override val commandAliases: List<String>
        get() = listOf(COMMAND_NAME)

    companion object {
        const val COMMAND_NAME = "get"
        const val ALL_FLAG = "-a"
        const val LOGS_FLAG = "-l"
        const val ENCS_FLAG = "-e"
    }
}